### DISCUSSION TITLE
RTK Query: Toast notification content contingent on multiple PUT responses

### DISCUSSION THREAD

**Discussion by nrl240** [reduxjs/redux-toolkit#2981] [Q&A] [answered]
Hello!

I'm toying with the idea of using RTK Query as I rebuild an app that leverages Redux Sagas. I've thoroughly read through the documentation, but I'm stumped on determining if RTK Query would be appropriate for the following use case:
- When updating settings for a given entity in the UI, multiple APIs must be hit.
- The toast notification that pops after the update is complete is dependent on all of the API responses.
  - For example, "Failed to update the following: • Enabled status • Products • Format," where three of the five `PUT` requests failed and their details were consolidated into a single toast notification.

Right now we handle this using Redux Sagas. We `yield all(requests)`, collect the failed responses within the saga's logic, and then dispatch the toast notification content/status.

Is this concept achievable with RTK Query alone?

**Note:** I see the documentation for [`createListenerMiddleware`](https://redux-toolkit.js.org/api/createListenerMiddleware). It looks like it each listener only listens for a single action. In my use case, I would need to listen for a series of actions after they're all complete (failed/succeeded). Apologies  in advance if I missed something in the documentation.

Thank you!

**Comment by markerikson** [comment, marked as answer]
That's definitely not achievable just with RTKQ, because RTKQ is _just_ about the data fetching aspect.

So yes, that's where listeners would come in, and listeners are much more flexible than "a single action" :)

Per https://redux-toolkit.js.org/api/createListenerMiddleware#startlistening , you can provide any one of the four options for deciding when an effect is kicked off:

- `type` or `actionCreator`: an exact match against the action type string
- `matcher`: a function that can be any combination of the [RTK "matcher" utilities](https://redux-toolkit.js.org/api/matching-utilities), such as `isAnyOf(action1, action2, action3)`
- `predicate`: any function with a signature of `(action, prevState, currentState) => boolean`, with arbitrary comparison logic inside

If you need to run logic in a sequence, you could kick off the effect when it sees the first qualifying action, then do `await take()` or `await condition()` inside to listen for the next steps:

- https://redux-toolkit.js.org/api/createListenerMiddleware#writing-async-workflows-with-conditions

**Reply by nrl240** [reply]
Beautiful! You are a machine, @markerikson! Thank you so much for your quick reply. I look forward to getting deeper into RTK and _finally_ RTKQ:)
